### PR TITLE
Fix linking Athena with VecGeom+CUDA

### LINKED ISSUE #1462: Apparent incompatibility with current ATLAS GPU environment
When running on `lxplus-gpu` we observed a crash within one of `Celeritas` test checking `Thrust` functionality.

With Celeritas using for example commit fe52da7e or commit aab88dd1365d313e3d13c482425426ae39601fcd 
and
```
export G4PATH=/cvmfs/atlas-nightlies.cern.ch/repo/sw/local/simulation/main_AthSimulation_x86_64-el9-gcc13-opt/Geant4
export ATLAS_LOCAL_ROOT_BASE="/cvmfs/atlas.cern.ch/repo/ATLASLocalRootBase"
source ${ATLAS_LOCAL_ROOT_BASE}/user/atlasLocalSetup.sh

asetup AthSimulation,local/simulation/main_AthSimulation_x86_64-el9-gcc13-opt,2023-10-15T1800

cmake -DVDT_ROOT=/cvmfs/atlas-nightlies.cern.ch/repo/sw/local/simulation/main_AthSimulation_x86_64-el9-gcc13-opt/sw/lcg/releases/LCG_106_ATLAS_13/vdt/0.4.4/x86_64-el9-gcc13-opt/ -DCELERITAS_USE_Geant4=OFF -DCELERITAS_BUILD_TESTS=ON -DCELERITAS_USE_CUDA=ON -DCELERITAS_USE_ROOT=ON -DCMAKE_CUDA_ARCHITECTURES=75 -G Ninja -DCMAKE_CUDA_FLAGS="-g -G" ../celeritas
```
we get a failure in `test/celeritas/track_TrackSort`

```
status: Celeritas core state initialization complete

CUDA Exception: Warp Illegal Address
The exception was triggered at PC 0x7fff6b4f9cd0 (agent_radix_sort_downsweep.cuh:293)

Thread 1 "track_TrackSort" received signal CUDA_EXCEPTION_14, Warp Illegal Address.
[Switching focus to CUDA kernel 4, grid 2307, block (0,0,0), thread (32,0,0), device 0, sm 0, warp 0, lane 0]
0x00007fff6b4f9a50 in cub::CUB_200302_750_NS::AgentRadixSortDownsweep<cub::CUB_200302_750_NS::AgentRadixSortDownsweepPolicy<512, 23, unsigned int, (cub::CUB_200302_750_NS::BlockLoadAlgorithm)3, (cub::CUB_200302_750_NS::CacheLoadModifier)0, (cub::CUB_200302_750_NS::RadixRankAlgorithm)2, (cub::CUB_200302_750_NS::BlockScanAlgorithm)2, 7, cub::CUB_200302_750_NS::RegBoundScaling<512, 23, unsigned int> >, false, unsigned int, unsigned int, unsigned int, cub::CUB_200302_750_NS::detail::identity_decomposer_t>::ScatterKeys<false> (this=0x7fffd7fffb58, twiddled_keys=..., relative_bin_offsets=..., ranks=..., valid_items=6240)
    at /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/agent/agent_radix_sort_downsweep.cuh:293
293	            temp_storage.keys_and_offsets.exchange_keys[ranks[ITEM]] = twiddled_keys[ITEM];
(cuda-gdb) p ranks
$2 = (@local int (& @local)[23]) <error reading variable: Error: read_local_memory(0, 0, 0, 0): failed to read local memory at address 0x7fffd7fff8f8 size 92, error=CUDBG_ERROR_INVALID_ADDRESS, error message=
>
(cuda-gdb) p twiddled_keys
$3 = (@local _ZN3cub17CUB_200302_750_NS23AgentRadixSortDownsweepINS0_29AgentRadixSortDownsweepPolicyILi512ELi23EjLNS0_18BlockLoadAlgorithmE3ELNS0_17CacheLoadModifierE0ELNS0_18RadixRankAlgorithmE2ELNS0_18BlockScanAlgorithmE2ELi7ENS0_15RegBoundScalingILi512ELi23EjEEEELb0EjjjNS0_6detail21identity_decomposer_tEE16bit_ordered_typeE (& @local)[23]) <error reading variable: Error: read_local_memory(0, 0, 0, 0): failed to read local memory at address 0x7fffd7fffaa4 size 92, error=CUDBG_ERROR_INVALID_ADDRESS, error message=
```

`compute-sanitize` reports
```
========= Invalid __global__ write of size 4 bytes
=========     at ScatterKeys<false>+0x323d0 in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/agent/agent_radix_sort_downsweep.cuh:293
=========     by thread (32,0,0) in block (0,0,0)
=========     Address 0x7f2a612882f8 is out of bounds
=========     and is 2391310585 bytes after the nearest allocation at 0x7f29d2a00000 of size 512 bytes
=========     Device Frame:ProcessTile<false>+0x33710 in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/agent/agent_radix_sort_downsweep.cuh:635
=========     Device Frame:ProcessRegion+0x363d0 in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/agent/agent_radix_sort_downsweep.cuh:820
=========     Device Frame:void cub::CUB_200302_750_NS::DeviceRadixSortDownsweepKernel<cub::CUB_200302_750_NS::DeviceRadixSortPolicy<unsigned int, unsigned int, unsigned int>::Policy900, (bool)0, (bool)0, unsigned int, unsign
ed int, unsigned int, cub::CUB_200302_750_NS::detail::identity_decomposer_t>(const T4 *, T4 *, const T5 *, T5 *, T6 *, T6, int, int, cub::CUB_200302_750_NS::GridEvenShare<T6>, T7)+0x36420 in /cvmfs/projects.cern.ch/lcg/releas
es/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/agent/agent_radix_sort_downsweep.cuh:339
=========     Saved host backtrace up to driver entry point at kernel launch time
=========     Host Frame: [0x2f167f]
=========                in /lib64/libcuda.so.1
=========     Host Frame: [0x15803]
=========                in /cvmfs/atlas-nightlies.cern.ch/repo/sw/local/simulation/main_AthSimulation_x86_64-el9-gcc13-opt/sw/lcg/releases/cuda/12.4.1-5bf34/x86_64-el9-gcc13-opt/lib64/libcudart.so.12
=========     Host Frame:cudaLaunchKernel [0x75230]
=========                in /cvmfs/atlas-nightlies.cern.ch/repo/sw/local/simulation/main_AthSimulation_x86_64-el9-gcc13-opt/sw/lcg/releases/cuda/12.4.1-5bf34/x86_64-el9-gcc13-opt/lib64/libcudart.so.12
=========     Host Frame:cudaError cudaLaunchKernel<char>(char const*, dim3, dim3, void**, unsigned long, CUstream_st*) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cuda_runtim
e.h:216 [0xd772fa]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:__device_stub__ZN3cub17CUB_200302_750_NS30DeviceRadixSortDownsweepKernelINS0_21DeviceRadixSortPolicyIjjjE9Policy900ELb0ELb0EjjjNS0_6detail21identity_decomposer_tEEEvPKT2_PS7_PKT3_PSB_PT4_SF_iiNS0_13Gr
idEvenShareISF_EET5_(unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, unsigned int*, unsigned int, int, int, cub::CUB_200302_750_NS::GridEvenShare<unsigned int>&, cub::CUB_200302_750_NS::detail::identit
y_decomposer_t&) in /tmp/pcanal/tmpxft_00233a7e_00000000-6_TrackSortUtils.cudafe1.stub.c:55 [0xd75a4d]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void cub::CUB_200302_750_NS::__wrapper__device_stub_DeviceRadixSortDownsweepKernel<cub::CUB_200302_750_NS::DeviceRadixSortPolicy<unsigned int, unsigned int, unsigned int>::Policy900, false, false, uns
igned int, unsigned int, unsigned int, cub::CUB_200302_750_NS::detail::identity_decomposer_t>(unsigned int const*&, unsigned int*&, unsigned int const*&, unsigned int*&, unsigned int*&, unsigned int&, int&, int&, cub::CUB_200
302_750_NS::GridEvenShare<unsigned int>&, cub::CUB_200302_750_NS::detail::identity_decomposer_t&) in /tmp/pcanal/tmpxft_00233a7e_00000000-6_TrackSortUtils.cudafe1.stub.c:58 [0xd75abe]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void cub::CUB_200302_750_NS::DeviceRadixSortDownsweepKernel<cub::CUB_200302_750_NS::DeviceRadixSortPolicy<unsigned int, unsigned int, unsigned int>::Policy900, false, false, unsigned int, unsigned int
, unsigned int, cub::CUB_200302_750_NS::detail::identity_decomposer_t>(unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, unsigned int*, unsigned int, int, int, cub::CUB_200302_750_NS::GridEvenShare<unsigned int>, cub::CUB_200302_750_NS::detail::identity_decomposer_t) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/device/dispatch/dispatch_radix_sort.cuh:303 [0xd792a8]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:cudaError thrust::THRUST_200302_750_NS::cuda_cub::launcher::triple_chevron::doit_host<void (*)(unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, unsigned int*, unsigned int, int, int, cub::CUB_200302_750_NS::GridEvenShare<unsigned int>, cub::CUB_200302_750_NS::detail::identity_decomposer_t), unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, unsigned int*, unsigned int, int, int, cub::CUB_200302_750_NS::GridEvenShare<unsigned int>, cub::CUB_200302_750_NS::detail::identity_decomposer_t>(void (*)(unsigned int const*, unsigned int*, unsigned int const*, unsigned int*, unsigned int*, unsigned int, int, int, cub::CUB_200302_750_NS::GridEvenShare<unsigned int>, cub::CUB_200302_750_NS::detail::identity_decomposer_t), unsigned int const* const&, unsigned int* const&, unsigned int const* const&, unsigned int* const&, unsigned int* const&, unsigned int const&, int const&, int const&, cub::CUB_200302_750_NS::GridEvenShare<unsigned int> const&, cub::CUB_200302_750_NS::detail::identity_decomposer_t const&) const in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/core/triple_chevron_launch.h:70 [0xd8df37]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:cudaError cub::CUB_200302_750_NS::DeviceRadixSort::SortPairs<unsigned int, unsigned int, int>(void*, unsigned long&, cub::CUB_200302_750_NS::DoubleBuffer<unsigned int>&, cub::CUB_200302_750_NS::DoubleBuffer<unsigned int>&, int, int, int, CUstream_st*) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/cub/device/device_radix_sort.cuh:806 [0xd82956]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:cudaError thrust::THRUST_200302_750_NS::cuda_cub::__radix_sort::dispatch<cuda::std::__4::integral_constant<bool, true>, thrust::THRUST_200302_750_NS::less<unsigned int> >::doit<unsigned int, unsigned int, long>(void*, unsigned long&, cub::CUB_200302_750_NS::DoubleBuffer<unsigned int>&, cub::CUB_200302_750_NS::DoubleBuffer<unsigned int>&, long, CUstream_st*) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/sort.h:279 [0xd7d01f]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void thrust::THRUST_200302_750_NS::cuda_cub::__radix_sort::radix_sort<cuda::std::__4::integral_constant<bool, true>, thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base>, unsigned int, unsigned int, long, thrust::THRUST_200302_750_NS::less<unsigned int> >(thrust::THRUST_200302_750_NS::cuda_cub::execution_policy<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base> >&, unsigned int*, unsigned int*, long, thrust::THRUST_200302_750_NS::less<unsigned int>) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/sort.h:367 [0xd7cb08]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:thrust::THRUST_200302_750_NS::cuda_cub::__smart_sort::enable_if_primitive_sort<unsigned int*, thrust::THRUST_200302_750_NS::less<unsigned int> >::type thrust::THRUST_200302_750_NS::cuda_cub::__smart_sort::smart_sort<cuda::std::__4::integral_constant<bool, true>, cuda::std::__4::integral_constant<bool, false>, thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base>, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>, thrust::THRUST_200302_750_NS::less<unsigned int> >(thrust::THRUST_200302_750_NS::cuda_cub::execution_policy<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base> >&, unsigned int*, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>, thrust::THRUST_200302_750_NS::less<unsigned int>) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/sort.h:461 [0xd7c674]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void thrust::THRUST_200302_750_NS::cuda_cub::sort_by_key<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base>, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>, thrust::THRUST_200302_750_NS::less<unsigned int> >(thrust::THRUST_200302_750_NS::cuda_cub::execution_policy<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base> >&, unsigned int*, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>, thrust::THRUST_200302_750_NS::less<unsigned int>) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/sort.h:551 [0xd7c166]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
[1:25](https://celeritasproject.slack.com/archives/DSUBD0FLM/p1729319158624659)
=========     Host Frame:void thrust::THRUST_200302_750_NS::cuda_cub::sort_by_key<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base>, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int> >(thrust::THRUST_200302_750_NS::cuda_cub::execution_policy<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base> >&, unsigned int*, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/system/cuda/detail/sort.h:621 [0xd7b15c]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void thrust::THRUST_200302_750_NS::sort_by_key<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base>, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int> >(thrust::THRUST_200302_750_NS::detail::execution_policy_base<thrust::THRUST_200302_750_NS::detail::execute_with_allocator<thrust::THRUST_200302_750_NS::mr::allocator<char, celeritas::AsyncMemoryResource<void*> >, thrust::THRUST_200302_750_NS::cuda_cub::execute_on_stream_nosync_base> > const&, unsigned int*, unsigned int*, thrust::THRUST_200302_750_NS::device_ptr<unsigned int>) in /cvmfs/projects.cern.ch/lcg/releases/cuda/12.4.1/x86_64-linux/targets/x86_64-linux/include/thrust/detail/sort.inl:102 [0xd79535]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void celeritas::detail::(anonymous namespace)::sort_impl<celeritas::OpaqueId<celeritas::ActionInterface, unsigned int>, unsigned int>(celeritas::Collection<unsigned int, (celeritas::Ownership)1, (celeritas::MemSpace)1, celeritas::OpaqueId<celeritas::Thread_, unsigned int> > const&, celeritas::ObserverPtr<celeritas::OpaqueId<celeritas::ActionInterface, unsigned int> const, (celeritas::MemSpace)1>, celeritas::OpaqueId<celer
itas::Stream_, unsigned int>) in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/track/detail/TrackSortUtils.cu:98 [0xd78310]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:celeritas::detail::sort_tracks(celeritas::CoreStateData<(celeritas::Ownership)1, (celeritas::MemSpace)1> const&, celeritas::TrackOrder) in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/track/detail/TrackSortUtils.cu:155 [0xd738e1]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:celeritas::SortTracksAction::step(celeritas::CoreParams const&, celeritas::CoreState<(celeritas::MemSpace)1>&) const in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/track/SortTracksAction.cc:121 [0xc2aa05]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:void celeritas::ActionSequence::step<(celeritas::MemSpace)1>(celeritas::CoreParams const&, celeritas::CoreState<(celeritas::MemSpace)1>&) in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/global/ActionSequence.cc:130 [0xae7110]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:celeritas::Stepper<(celeritas::MemSpace)1>::operator()() in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/global/Stepper.cc:130 [0xb21ef9]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:celeritas::Stepper<(celeritas::MemSpace)1>::operator()(celeritas::Span<celeritas::Primary const, 18446744073709551615ul>) in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/src/celeritas/global/Stepper.cc:166 [0xb22272]
=========                in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/build-celeritas-cuda-debug/lib64/libceleritas.so
=========     Host Frame:celeritas::test::TestTrackSortActionIdEm3Stepper_device_is_sorted_Test::TestBody() in /afs/cern.ch/user/p/pcanal/atlas-on-eos/celeritas/celeritas/test/celeritas/track/TrackSort.test.cc:394 [0x43529]
```

### PR DESCRIPTION
**Pull request by pcanal**
`libvecgeomcuda.[a|so]` is the final library and it includes the result
of nvlink.  It is not appropriate to use as the intermediary link library.

VecGeom also never produce a shared 'middle' library, so we just always
need to use`libvecgeomcuda.[a|so]` is the final library and it includes the result
of nvlink.  It is not appropriate to use as the intermediary link library.

VecGeom also never produce a shared 'middle' library, so we just always
need to use `libvecgeomcuda_static.a`

Other fixes includes:

* removing spurrious dependencies
* adding support for private linking
* fix handling of some internals when seeing static libraries.

### CODE REVIEW

**Review by sethrj — CHANGES_REQUESTED** [review]
Currently fails for me on a `+shared~geant4+vecgeom ^vecgeom+shared` build on wildstyle:
```
[906/1015] Linking CXX executable bin/orange-update
FAILED: bin/orange-update
: && /usr/bin/c++ -Wall -Wextra -pedantic -fdiagnostics-color=always -O3 -DNDEBUG  app/CMakeFiles/orange-update.dir/orange-update.cc.o -o bin/orange-update  -Wl,-rpath,/home/s3j/Code/celeritas/build-temp-nog4/lib64:/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64:  lib64/liborange_final.so  lib64/liborange.so  lib64/libcorecel.so  -Wl,-rpath-link,/home/s3j/Code/celeritas/build-temp-nog4/lib64:/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64 && :
/projects/spack/opt/spack/gcc-8.5.0/vecgeom/ykhz7jg/lib64/libvecgeom.so: undefined reference to `void vecgeom::cxx::DevicePtr<vecgeom::cuda::SIMDSpecializedVolImplHelper<vecgeom::cuda::EllipticalTubeImplementation, -1, -1> >::Construct<vecgeom::cxx::DevicePtr<vecgeom::cuda::LogicalVolume>, vecgeom::cxx::DevicePtr<vecgeom::cuda::Transformation3D>, unsigned int, int, int>(vecgeom::cxx::DevicePtr<vecgeom::cuda::LogicalVolume>, vecgeom::cxx::DevicePtr<vecgeom::cuda::Transformation3D>, unsigned int, int, int) const'
```

**Comment by pcanal** [comment]
The failure is reproduced locally.  The issue is that RdcCuda does not yet support the case of a `final` shared library with a `static` middle/object library (up to now they have always been the same type).  I am exploring 2 options.  Either update VecGeom to produce a 3rd library (the middle/object library) when building shared libraries or updated RdcCuda cmake fragment to handle the case.

**Comment by pcanal** [comment]
The only failure left is:
```
/usr/bin/ld: final link failed: No space left on device
```

**Comment by sethrj** [comment]
Um how much duplication is in these?

**Comment by pcanal** [comment]
> Um how much duplication is in these?

Well ... now we use vecgeom_static must be copied over I guess ... It can be fixed by changing the VecGeom build.

**Comment by sethrj** [comment]
So if we have our main libraries (3 of which indirectly link against vecgeom), 4 celeritas tests, and 5 apps, will we have 12 copies of all the vecgeom device symbols? Or 24? Or much fewer?

**Comment by pcanal** [comment]
The problem really comes from the executable with 179 of them being more than 30M ... I have to fix this ....

**Comment by pcanal** [comment]
Humm .. actually probably can't do anything about it short of having a non-final but shared `libvecgeomcuda` library.

For comparison the size of executable in the static-static build is 100MB !!!

**Comment by sethrj** [comment]
@pcanal Here are the sizes of the different shared+shared builds on wildstyle. The build directory is now 2-3× larger...

Screenshot 2024-11-08 at 08 05 38

**Comment by sethrj** [comment]
@pcanal Are the file sizes smaller with this build? Will it require the pending vecgeom MR to go in?

**Comment by pcanal** [comment]
The size should be back to normal.  This PR does not require the VecGeom update but will take advantage of it as soon as it is available.

**Comment by pcanal** [comment]
@sethrj Do you know why the docker tests are being skipped?

**Comment by sethrj** [comment]
@pcanal To reduce resource use they're skipped in draft mode. I'll take a look in a minute (build on wildstyle) and then the next push should trigger docker as well.

**Comment by sethrj** [comment]
percentages

Hey, with the changes the reldeb sizes are even smaller! Thanks @pcanal

File: cmake/FindVecGeom.cmake [hunk 1]
```diff
@@ -26,21 +26,26 @@ if(VecGeom_FOUND AND TARGET VecGeom::vecgeomcuda)
   get_target_property(_vecgeom_lib_type VecGeom::vecgeom TYPE)
   if (_vecgeom_lib_type STREQUAL "STATIC_LIBRARY")
      set(_vecgeom_cuda_runtime "Static")
+     set(_vecgeom_middle_library_suffix "_static")
   else()
      set(_vecgeom_cuda_runtime "Shared")
+     set(_vecgeom_middle_library_suffic "")
```
Comment (sethrj) [inline cmake/FindVecGeom.cmake]:
```suggestion
     set(_vecgeom_middle_library_suffix "")
```

File: cmake/FindVecGeom.cmake [hunk 2]
```diff
@@ -51,15 +56,17 @@ if(VecGeom_FOUND AND TARGET VecGeom::vecgeomcuda)
...
-  foreach(_lib VecGeom::vecgeom VecGeom::vecgeomcuda
-      VecGeom::vecgeomcuda_static)
-    set_target_properties(${_lib} PROPERTIES
-      CUDA_RDC_STATIC_LIBRARY VecGeom::vecgeomcuda_static
-      CUDA_RDC_MIDDLE_LIBRARY VecGeom::vecgeomcuda
-      CUDA_RDC_FINAL_LIBRARY VecGeom::vecgeomcuda
-    )
-  endforeach()
+  if(NOT _vecgeom_lib_rdc_final)
+    # Inform celeritas_add_library code
```
Comment (sethrj) [inline cmake/FindVecGeom.cmake]: Should be `cuda_rdc_add_library` now?

**Review by sethrj — APPROVED** [review]
Great  work!

**Review by esseivaju — APPROVED** [review]
Looks good! Tested within Athena and everything's working

**Pull request merged**